Sherwin/pk game

sherrpass opened the pull request:
## Deals with:
#58 #57 #56 #54 (Need help changing fonts)

## Completed:
- updates and sync pk winning rate
- Implemented PKGame, with matchmaking.
- Also added devLogin page for easy login between dev accounts

## TODO:
-  [x] Leave game midway responsibly functionality 
-  [x] Updating of win rate after game
- [ ] Add timer
- [ ] PK button works on all book cards
- [ ] improve efficiency of firebase query for PKGame
- [ ] Add PK words summary popup in outcome page.

sherrpass committed:
implement matchfinding

sherrpass committed:
implement local version of PKGame

sherrpass committed:
implement online PKgame quiz

sherrpass committed:
implement PKgame overview page

sherrpass committed:
Merge branch 'main' of https://github.com/LingoClash/LingoClash into sherwin/PKGame

sherrpass committed:
refactor PKGame engine to improve code quality

sherrpass committed:
add devLogin page for easy login, refactor to work with revamp ui

sherrpass committed:
fix build issues

sherrpass committed:
Merge branch 'main' of https://github.com/LingoClash/LingoClash into sherwin/PKGame

sherrpass committed:
fix production build error

sherrpass committed:
fix production build issues

sherrpass committed:
remove duplicate files

sherrpass committed:
remove more duplicate files

sherrpass committed:
tidy up storyboard file structure

sherrpass committed:
remove redundant storyboard

sherrpass commented on LingoClash/LingoClash/Model/DataManagers/DataManagerResources.swift:
```diff
@@ -0,0 +1,20 @@
+//
+//  Created by Sherwin Poh on 14/4/22.
+//
+
```
created this because i needed to access the pkGames and queueEntries collection name in my PKGameMoveUpdater and PKGameMatchFinder. Decided to do it for the rest as well to be consistent

sherrpass commented on LingoClash/LingoClash/DataProviders/DataProviderTypes.swift:
```diff
@@ -13,11 +13,16 @@ protocol Record: Codable {
 }
 
+extension Record {
+    // Local creation of Records to be pushed to DataProvider where id is useless
```
As described this placeHolderId is used for the creation of Records locally to then be created/ used for update in DataManager. Id cannot be blank

sherrpass commented on LingoClash/LingoClash/DataProviders/Firebase/FirebaseDataProvider.swift:
```diff
@@ -54,9 +54,17 @@ class FirebaseDataProvider: DataProvider {
         
         return Promise { seal in
-            db.collection(resource).getDocuments { (querySnapshot, error) in
+            var collectionQuery: Query = db.collection(resource)
```
Added the filter and sort mechanics here as well

sherrpass commented on LingoClash/LingoClash/DataProviders/Firebase/FirebaseDataProvider.swift:
```diff
@@ -222,13 +231,32 @@ class FirebaseDataProvider: DataProvider {
                     }
-                    // TODO: Modify the id to be ref.documentID
-                    return seal.fulfill(CreateResult(data: params.data))
+                }
```
accomplished todo

sherrpass committed:
remove redundant files

sherrpass commented on LingoClash/LingoClash/Model/Games/PKGameMove.swift:
```diff
@@ -0,0 +1,29 @@
+    
+    // TODO: Possibly add a constant to represent a placeholder
+    init(question: Question, player: Profile, isCorrect: Bool, timeTaken: Double, id: Identifier = "") {
+        self.id = "Placeholder"
```
Need advice on this: @CrownKira / however reviews this. I need to create a PKGameMove locally, what should the id be?

CrownKira replied:
Ok i will look into this. Perhaps we can make the id an optional field

sherrpass commented on LingoClash/LingoClash/Model/Questions/QuestionContructors/MatchVocabToDefinitionQuestionConstructor.swift:
```diff
@@ -10,12 +10,13 @@ struct MatchVocabToDefinitionQuestionConstructor: QuestionContructor {
-        let options = [vocabsTested.map { $0.word }.shuffled(), vocabsTested.map { $0.definition }.shuffled()]
-        let answer = Set(vocabsTested.map { [$0.word, $0.definition] })
-        return MatchVocabToDefinitionQuestion(context: "Match the Vocabulary to its definition.", options: options, answer: answer, vocabsTested: vocabsTested)
+//        guard vocabsTested.count == vocabsTestedCount, otherVocabs.count == otherVocabsCount else {
```
commented out till actually implemented.

sherrpass commented on LingoClash/LingoClash/Model/Questions/QuestionTypes/QuestionWrapper.swift:
```diff
@@ -0,0 +1,45 @@
+//
+
+struct QuestionWrapper: Codable {
+    let question: Question
```
This is a wrapper to enable questions to be encodable and decodable. Unfortunately, cannot find an easier way out of this.

sherrpass commented on LingoClash/LingoClash/Model/Vocab.swift:
```diff
@@ -14,6 +14,14 @@ struct Vocab {
     let sentenceDefinition: String
     let pronunciationText: String
+    init(id: Identifier = "0" , word: String, definition: String, sentence: String, sentenceDefinition: String, pronunciationText: String) {
+        self.id = id
```
Same issue as above, what should id be?

sherrpass commented on LingoClash/LingoClash/PKGameView/PKGameLobby/FirebasePKGameMatchFinder.swift:
```diff
@@ -0,0 +1,186 @@
+        }.then { [self] pkGameData in
+            self.pkGameDataManager.createPKGame(pkGameData: pkGameData)
+        }.then { [self] newGame -> Promise<PKGame> in
+            let batch = self.db.batch()
+            for id in queueEntries.map({ $0.id }) {
```
TODO: refactor this to be in dataManager layer for abstraction to be consistent with rest.

sherrpass committed:
ensure matchfinding cancels responsibly, implement  PK again button and draw, match ui with theme

CrownKira committed:
Remove comments in appconfigs

CrownKira committed:
Merge branch 'sherwin/PKGame' of https://github.com/LingoClash/LingoClash into sherwin/PKGame

CrownKira committed:
Update PKGame UI to match main theme

CrownKira committed:
Remove extra references in PKGame storyboard

sherrpass committed:
change ui colour to fit theme

sherrpass committed:
refactor FirebasePKGameMoveUpdater to clean up listeners, add PKGame leave primitive functionality

sherrpass committed:
add rename GameMoveUpdater and listener, clean up question animations, prevent double selection

sherrpass committed:
add forfeitted players on pkGames firebase only

sherrpass committed:
add listener to PKGame and for forfeitted Players

sherrpass committed:
implement updating for forfeiting, pk_game_outcomes

sherrpass committed:
complete implementing forfeit, records pkgame player outcome in backend

sherrpass committed:
implement pk rate to sync with backend

CrownKira reviewed (changes requested):
Hi, i just tested the app. Here's are some bugs or issues that i think should be addressed before it is merged: 

- [ ] Vocabs Learnt not updated after learning a lesson 

- [ ] Lesson stars not updated after learning a lesson 
- [ ] Progress (in the card) not updated after learning a lesson 
(might be due to not refreshing the UI on viewDidLoad)
- [x] Learning page not working after this PR 
(Completed page not working also) 
- [ ] Perhaps all print statements should be replaced with Log statements.
- [x] App crashes when vocab count is less than 10; when click on start learning button 
<img width="500" alt="Screenshot 2022-04-16 at 3 24 37 AM" src="https://user-images.githubusercontent.com/24221801/163623537-7f6d703c-d0b0-4f11-8e23-1ea83b6f56b9.png">

 - [x] I have added translation field to vocabData. Do you mind adding it to the vocab learning page. Put the translation under the example sentence (secondary color) 
 
- [x] Encountered an error here; when i click on Lesson 9 of current book
<img width="500" alt="Screenshot 2022-04-16 at 3 47 49 AM" src="https://user-images.githubusercontent.com/24221801/163625462-7174335a-be77-441d-abb6-ed4bf2749ef2.png">

- [x] Should wrap long text; increase button height  
<img width="500" alt="Screenshot 2022-04-16 at 3 52 38 AM" src="https://user-images.githubusercontent.com/24221801/163625867-b755190a-ebd3-4e51-83bb-bf9123c14393.png">

- [ ] Perhaps remove the arena tab if you don't find it useful 

- [ ] Clear all linting warnings regarding your features

CrownKira committed:
Fix linting warnings and errors

CrownKira committed:
Parse korean JSON and add it to sample data

CrownKira committed:
Parse korean vocab despite empty definition

sherrpass committed:
fix resizing issues with lesson overview and character cutoff in lesson quiz

sherrpass committed:
Merge branch 'sherwin/PKGame' of https://github.com/LingoClash/LingoClash into sherwin/PKGame

ailing35 committed:
Fix Firebase data provider sorting bug

sherrpass committed:
work in progress for adding vocabsLearnt update

sherrpass committed:
Merge branch 'sherwin/PKGame' of https://github.com/LingoClash/LingoClash into sherwin/PKGame

sherrpass committed:
updates vocabsLearnt and stars for completed lessons with profileLessons created

sherrpass commented:
@CrownKira Not sure why i cannot reply to your comment. So my response are below

Hi, i just tested the app. Here's are some bugs or issues that i think should be addressed before it is merged:

 Vocabs Learnt not updated after learning a lesson
- Will now be updated IF profileLesson and profileBook is created when a new book is added. @ailing35 Does it currently?
 
Lesson stars not updated after learning a lesson
- It now will be updated in the backend again if profileLesson and book is created ....
- TODO: Update UI when this occurs.

 Progress (in the card) not updated after learning a lesson
(might be due to not refreshing the UI on viewDidLoad)

 Learning page not working after this PR
(Completed page not working also)
- fixed by kevin and aiing

 Perhaps all print statements should be replaced with Log statements.

 App crashes when vocab count is less than 10; when click on start learning button
- cannot recreate.

 I have added translation field to vocabData. Do you mind adding it to the vocab learning page. Put the translation under the example sentence (secondary color)
- THere already is a sentenceDefinition field in vocab class and a field on the storybaord and viewcontroller for this.

 Encountered an error here; when i click on Lesson 9 of current book
- Fixed, it was likely because a vocab length was too long which lead to a UICOllection layout issue.

 Should wrap long text; increase button height
- decreased font size slightly, added wrap up till two lines, any more will be truncated neatly.

 Perhaps remove the arena tab if you don't find it useful
- May use it in future, thanks for it!

 Clear all linting warnings regarding your features
- Will do springcleaning in future PR.

sherrpass committed:
Merge branch 'main' of https://github.com/LingoClash/LingoClash into sherwin/PKGame

CrownKira committed:
Add model details to dataprovider error log

CrownKira committed:
Add pk winning rate to profile data

CrownKira committed:
Adjust green to be brighter

CrownKira committed:
Fix edit profile form

CrownKira committed:
Some touchup

CrownKira reviewed (approved):
LGTM!
